# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/engyusufayman06/ml-internship-2026/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [ ]:
## 1. My Lane

I chose **Refresh / Content Opportunity Scoring**.

My goal is to identify which pages should be reviewed first for a possible content refresh, improvement, or monitoring.

I chose this lane because the starter dataset already contains useful signals such as impressions, clicks, CTR, average position, content age, freshness, word count, engagement, and trend direction. The Week 1 starter pipeline also showed that a learned ranking method can outperform a simple hand-written rule on the starter dataset.

This makes the lane a good fit for testing whether observable search and content signals can help prioritize limited review capacity.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [ ]:
## 2. The Question

### Research Question

**Which pages should be reviewed first for a potential content refresh or other improvement?**

### Decision

The decision is to determine which pages deserve higher priority for human review based on observable evidence from search and content performance data.

### Action

A reviewer could use the ranked output to decide whether a page should be:

- refreshed or updated,
- expanded,
- monitored,
- or potentially left unchanged.

The system is intended to prioritize pages for review rather than automatically decide what action should be taken.

### Cost of a Wrong Recommendation

A false positive could cause the team to spend limited time reviewing or improving a page that does not have a meaningful opportunity.

A false negative could cause an important page to be overlooked, potentially missing an opportunity to improve its search performance.

Because of this trade-off, the ranking should be evaluated using metrics such as Precision@K and should remain explainable to a human reviewer.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [3]:
import os
import sys
import subprocess

REPO_URL = "https://github.com/engyusufayman06/ml-internship-2026"
REPO_DIR = "ml-internship-2026"

if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

os.chdir(REPO_DIR)

print("Working directory:", os.getcwd())

Working directory: /content/ml-internship-2026


In [4]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Number of pages:", len(df))

print(
    "Correlation between search_volume and impressions_90d:",
    round(df["search_volume"].corr(df["impressions_90d"]), 3)
)

print(
    "Declining page rate:",
    round(
        df["trend_direction"].str.lower().eq("down").mean(),
        3
    )
)

Number of pages: 30000
Correlation between search_volume and impressions_90d: 0.001
Declining page rate: 0.542


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [5]:
## 4. Careful Words: What I Can and Can't Claim

### What I Can Claim

I can say that the starter dataset shows associations between observable search/content signals and the provided labels.

I can also say that, on the starter dataset and under its validation setup, the learned random forest ranking achieved higher Precision@50 than the hand-written baseline.

The results suggest that learned ranking may be useful for prioritizing pages for human review.

### What I Cannot Claim

I cannot claim that a particular signal causes a page to decline.

I cannot claim that refreshing a page will definitely improve its traffic or rankings.

I cannot claim that the model has discovered a Google ranking algorithm factor.

I also cannot treat the starter `trend_direction == "down"` label as a perfect definition of a real future decline. It is a proxy label based on the current data window.

Therefore, my results will be described as observed or directional rather than causal or guaranteed.

SyntaxError: invalid syntax (3650573637.py, line 5)

### What I observed

The starter dataset contains **30,000 pages**.

The correlation between `search_volume` and `impressions_90d` is **0.001**, which is almost zero in this dataset. This suggests that keyword search volume alone is not a strong indicator of the impressions a page actually receives.

About **54.2%** of the pages are labeled as declining in the starter dataset.

The earlier starter pipeline also showed that the hand-written baseline achieved **0.240 Precision@50**, while the random forest achieved **0.740 Precision@50** on the evaluated starter data. This means that approximately 12 of the top 50 baseline recommendations and 37 of the top 50 random-forest recommendations were declining according to the starter label.

These observations make refresh opportunity scoring worth investigating further, while recognizing that the starter label is only a proxy for the real business outcome.

## 5. Self-Check

- **Problem:** Prioritize pages that may deserve content review.
- **Research question:** Which pages should be reviewed first for a potential refresh or improvement?
- **Unit of analysis:** One content page.
- **Decision:** Which pages should receive higher review priority?
- **Action:** A human reviewer investigates the highest-ranked pages and decides whether to refresh, expand, monitor, or leave them unchanged.
- **Cost of a wrong call:** False positives waste review time, while false negatives may cause useful opportunities to be missed.
- **Data:** Observable search and content performance signals from the anonymized starter dataset.
- **Baseline:** A simple hand-written ranking rule.
- **ML approach:** Test interpretable models and compare them with the baseline.
- **Evaluation:** Precision@K and other ranking metrics.
- **Leakage control:** Do not use information derived from the outcome or future target period as features.
- **Causality:** The project supports prioritization and decision-making; it does not prove that a content change causes recovery.
- **Public safety:** Only anonymized IDs and aggregated or safe metrics will be used in public outputs.